src/extension/components/PopupWindow.re

```diff
@@ -0,0 +1,143 @@
+open Core;
+
+let getLangList conversionType =>
+  switch (conversionType |> Js.String.split "to") {
+  | [| inLang, outLang |] => (Some inLang, Some outLang)
+  | _ => (None, None)
+};
+
+module PopupWindow = {
+  include ReactRe.Component.Stateful.InstanceVars;
+  let name = "PopupWindow";
+
+  type props = {
+    initialText: string
+  };
+  type state = {
+    copy: bool,
+    in_: string,
+    out: string,
+    inLang: option string,
+    outLang: option string
+  };
+  type instanceVars = {
+    mutable _in: option ReactRe.reactRef
+  };
+
+  let getInstanceVars () => {
+    _in: None
+  };
+
+  let getInitialState props => {
+    copy: false,
+    in_: "",
+    out: "",
+    inLang: None,
+    outLang: None
+  };
+
+  let copy { state, updater } () => {
+    Util.setTimeout (updater (fun { state } () => Some { ...state, copy: false })) 2500;
+    Some { ...state, copy: true }
+  };
+
+  let open_ { state } e => {
+    Chrome.Tabs.create { "url": ("popup.html#" ^ (Util.btoa state.in_)) };
+    None
+  };
+
+  let refmt { state, updater } value => {
+    open Background;
+    /* this isn't guaranteed to be sync or speedy, so
+     * don't set this.state.in here, since it could cause lag.
+     */
+    Chrome.Runtime.sendMessage
+      { in_: value }
+      (fun { out: (conversionType, out) } => {
+        let (inLang, outLang) = getLangList conversionType;
+        updater (fun { state } () => Some { ...state, out, inLang, outLang }) ()
+      });
+
+    Chrome.Storage.Local.set { "latestRefmtString": value };
+    Some { ...state, in_: value }
+  };
+
+  let componentDidMount { props, instanceVars, updater } => {
+    updater refmt props.initialText;
+
+    switch instanceVars._in {
+      | None => ()
+      | Some ref => {
+        CM.focus ref;
+        CM.execCommand ref "selectAll";
+      }
+    };
+    None
+  };
+
+  let render { props, state, updater, refSetter } => {
+    /* extra cautious here because of the async on _refmt
+     * sometimes typing can flash the wrong lang very quickly
+     */
+    let hasConverted = not (Str.isEmpty state.in_) && not (Str.isEmpty state.out);
+    let inTitle = hasConverted ? state.inLang |> (Option.map_or (fun lang => "In (" ^ lang ^ ")") "In") : "In";
+    let outTitle = hasConverted ? state.outLang |> (Option.map_or (fun lang => "Out (" ^ lang ^ ")") "Out") : "Out";
```

Comment:
You don't need the hasConverted check here, right? Since inLang and outLang are only true when hasConverted is also true, the single map_or should be sufficient.